### ISSUE TITLE
Lambda container updates on every reconsilation loop

### ISSUE THREAD

**Issue by candonov** [crossplane-contrib/provider-upjet-aws#692]
### What happened?

When installing lambda from a container image, every reconciliation loop updates the lambda. This only happens when the code is in a container image and NOT when the code is in an s3 tgz file.

### How can we reproduce it?

It can be reproduced by following these [instructions](https://github.com/awslabs/crossplane-on-eks/tree/main/examples/upbound-aws-provider/composite-resources/serverless-examples/sqs-lambda-s3). XRD and compositions are located [here](https://github.com/awslabs/crossplane-on-eks/tree/main/compositions/upbound-aws-provider/lambda). The issue occurs when using the `container.yaml` Composition.
Here is the `terraform plan` output from the upbound-aws-provider pod relevant workspace:
```
/tmp/56c45f56-2abc-47e9-aaf7-06a079b1faa3 $ terraform plan
aws_lambda_function.sqs-lambda-s3-test-6hsqt-processor: Refreshing state... [id=sqs-lambda-s3-test-6hsqt-processor]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # aws_lambda_function.sqs-lambda-s3-test-6hsqt-processor will be updated in-place
  ~ resource "aws_lambda_function" "sqs-lambda-s3-test-6hsqt-processor" {
        id                             = "sqs-lambda-s3-test-6hsqt-processor"
      ~ last_modified                  = "2023-05-04T17:20:47.000+0000" -> (known after apply)
      ~ source_code_hash               = "81f9eebaf9bcaf49f9ca7577fa80f2f36923ca34e88a1e0e1900e519b622e20d" -> "f5c00f06bd73f7141404c16690f50afa8b30fc33933feeed20f2ee25f2a8bd47"
        tags                           = {
            "crossplane-kind"           = "function.lambda.aws.upbound.io"
            "crossplane-name"           = "sqs-lambda-s3-test-6hsqt-processor"
            "crossplane-providerconfig" = "aws-provider-config"
        }
        # (17 unchanged attributes hidden)
        # (3 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
```
Per Terraform docs [`source_code_hash`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/lambda_function#source_code_hash) is optional and should be used only for `s3_key` option and not with `imageUri`. 

### What environment did it happen in?

* Crossplane Version: 1.11.2
* Provider Version: 0.32.1
* Kubernetes Version:  `Client Version: v1.27.1   Server Version: v1.25.6-eks-48e63af`
* Kubernetes Distribution:  EKS

**Comment by svscheg** [comment]
Issue reproduced on latest master:
TERRAFORM_PROVIDER_VERSION := 4.56.0
Provider version: v0.34.0

**Comment by candonov** [comment]
We need a way in the upjet implementation to conditionally skip in `LateInitializer` the `source_code_hash` if `imageUri` is provided, not `s3_key` or `filename`. 
I removed by hand the `source_code_hash` from the crossplane managed resource, the first reconcile loop it came back, and then the recourse is ready and synced trough.
cc: @ulucinar

**Comment by candonov** [comment]
related tf issue: https://github.com/hashicorp/terraform-provider-aws/issues/29921

**Comment by turkenh** [comment]
What is the exact behavior you're observing with the Lambda Function Managed resource? Is it not becoming synced/ready? Is it getting updates in every reconciles?

**Comment by candonov** [comment]
> 

The latter, it is getting updated on every reconcile loop, even though nothing changed.
<img width="1493" alt="Screenshot 2023-06-09 at 11 28 11 AM" src="https://github.com/upbound/provider-aws/assets/25967713/398b8b04-c32c-4783-b962-e7a628649832">
<img width="961" alt="Screenshot 2023-06-09 at 11 31 54 AM" src="https://github.com/upbound/provider-aws/assets/25967713/83680654-40ee-4e46-b532-0c0190ae33cd">
```
k get events |grep function
2m5s        Normal    UpdatedExternalResource   function/sqs-lambda-s3-test-r5w5s-processor                        Successfully requested update of external resource
```

**Comment by yebolenko** [comment]
Hi guys
We have an opposite behavior.
We use xpkg.upbound.io/upbound/provider-aws:v0.38.0
Lambda source is image (not zip)
When our pipeline updates lambda's image, we don't see that AWS Lambda is actually takes the latest image version (from ECR).
I need to delete AWS Lambda function manually and only after this it gets the right image.
We also observe an additional issue: in argocd we see that lambda function is sync, but actually lambda function has not been created yet.
Please advise what is wrong with image update.

**Comment by yebolenko** [comment]
any update?

**Comment by JohnLBevan** [comment]
bump

**Comment by jeanduplessis** [comment]
Has anyone run into this problem on the latest version of the provider? I.e. v1.3.0+?
Since this issue was triaged a long time ago, I'll mark it to be triaged again on a more recent version of the provider as a lot has changed since then.

**Comment by JohnLBevan** [comment]
@jeanduplessis Yes... there's a duplicate of this issue (later issue number, but same repo / same problem) on which the latest post confirms this issue is being produced on v1.30.0: https://github.com/crossplane-contrib/provider-upjet-aws/issues/1027#issuecomment-2034304044

**Comment by jeanduplessis** [comment]
Thanks @JohnLBevan.

I am going close this issue in favor of #1027.
We won't be able to look at it this week, but will prioritize it for the week of April 15th.
